### PR TITLE
Add streams and comms as optional arguments

### PR DESCRIPTION
**Pull request by colesbury**
Adds streams and comms as optional arguments to the NCCL calls in
torch.cuda.nccl. Also exposes ncclUniqueId and ncclCommInitRank for
multi-process mode.

Moves Py_RETURN_NONE statements after the GIL is re-acquired.

### CODE REVIEW

File: torch/csrc/cuda/nccl.cpp [hunk 1]
```diff
@@ -35,30 +38,35 @@ struct NcclCommList {
...
 
 // accesses to this object have to be guarded by THC's CudaFreeMutex
-std::unordered_map<std::string, NcclCommList > _communicators;
+std::unordered_map<std::string, NcclCommList> _communicators;
 
-static ncclComm_t* _get_communicator(std::vector<at::Tensor>& inputs) {
-  int ndevices = inputs.size();
+static ArrayRef<ncclComm_t> _get_communicators(TensorList inputs, ArrayRef<ncclComm_t> user_comms) {
+  if (user_comms.size() == inputs.size()) {
+    return user_comms;
```
Comment (apaszke) [inline torch/csrc/cuda/nccl.cpp]: I'd say that doing no checks for `user_comms` and silently falling back to regular operation if they don't match the inputs is a bit weird. You give the comms, but you can't even be sure if they are used. This should probably be an error.

File: torch/csrc/cuda/nccl.cpp [hunk 2]
```diff
@@ -121,232 +129,330 @@ static void _check_inputs(std::vector<at::Tensor> &inputs, std::vector<at::Tenso
...
+    auto seq = THPObjectPtr(PySequence_Fast(obj, "comm is not a sequence"));
+    if (!seq) throw python_error();
+    auto size = PySequence_Fast_GET_SIZE(seq.get());
+    comms = std::vector<ncclComm_t>(size);
+    for (int64_t i = 0; i < size; i++) {
+      comms[i] = unpack_nccl_comm(PySequence_Fast_GET_ITEM(seq.get(), i));
+    }
+  }
+  if (comms.size() != size) {
+    throw std::runtime_error("number of communicators is not equal to number of inputs");
```
Comment (apaszke) [inline torch/csrc/cuda/nccl.cpp]: Uh now I see that the condition in `_get_communicators` will always be satisfied, but I still think that doing it like this:

```cpp
auto comms = user_comms.empty() ? _get_communicator(inputs) : user_comms;
```

would be much clearer.

File: torch/csrc/utils/auto_gil.h [hunk 3]
```diff
@@ -25,3 +25,10 @@ struct AutoNoGIL {
 
   PyThreadState* save;
 };
+
+// Runs the function without the GIL
+template<typename F>
+inline void with_no_gil(F f) {
```
Comment (apaszke) [inline torch/csrc/utils/auto_gil.h]: that's a nice one

File: torch/csrc/cuda/nccl.cpp [hunk 4]
```diff
@@ -121,232 +129,330 @@ static void _check_inputs(std::vector<at::Tensor> &inputs, std::vector<at::Tenso
...
 
+static void _start_group() {
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+  CHECK(ncclGroupStart());
+#endif
+}
+
+static void _end_group() {
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+  CHECK(ncclGroupEnd());
```
Comment (apaszke) [inline torch/csrc/cuda/nccl.cpp]: Not sure what are the semantics of `ncclGroupEnd`, but if it should be called even if one of the collective calls fails, then we should make a scope guard with it.

**Review by apaszke — APPROVED** [review]
LGTM. I posted a few comments, but they are minor.

File: torch/csrc/cuda/nccl.cpp [hunk 5]
```diff
@@ -70,7 +80,7 @@ static ArrayRef<ncclComm_t> _get_communicators(TensorList inputs, ArrayRef<ncclC
   }
 }
 
-static void _check_inputs(std::vector<at::Tensor> &inputs, std::vector<at::Tensor> &outputs, int input_multiplier, int output_multiplier) {
+static void _check_inputs(TensorList inputs, TensorList outputs, int input_multiplier, int output_multiplier) {
```
Comment (apaszke) [inline torch/csrc/cuda/nccl.cpp]: They used to be references and now are passed by value. Is that expected?

Comment (colesbury) [reply torch/csrc/cuda/nccl.cpp]: `TensorList` is intended to be passed by value (just a pointer and size). It is typedef of `ArrayRef<Tensor>`.

Comment (apaszke) [reply torch/csrc/cuda/nccl.cpp]: Right, I always forget that ATen typedefs are `ArrayRef`s.

**Review by apaszke — APPROVED** [review]

File: torch/csrc/cuda/nccl.cpp (hunk 5 above)
Comment (ezyang) [reply torch/csrc/cuda/nccl.cpp]: But remember that `tensor_list` is a vector of tensors ;) (our naming convention is awful)

**Pull request merged**